Source/EasyNetQ/IEventBus.cs

```diff
@@ -1,102 +1,106 @@
+        /// <typeparam name="TEvent">The event type</typeparam>
+        /// <returns>Unsubscription disposable</returns>
+        IDisposable Subscribe<TEvent>(Action<TEvent> handler);
     }
 
-    public class EventBus : IEventBus
+    /// <inheritdoc />
+    public sealed class EventBus : IEventBus
     {
-        private class Handlers
-        {
-            private readonly object internalHandlersLock = new object();
-            private readonly List<object> internalHandlers;
-
-            public Handlers()
-            {
-                internalHandlers = new List<object>();
-            }
-
-            public Handlers(params object[] handlers)
-            {
-                internalHandlers = new List<object>(handlers);
-            }
-
-            public void Add(object handler)
-            {
-                lock (internalHandlersLock)
-                    internalHandlers.Add(handler);
-            }
-
-            public void Remove(object handler)
-            {
-                lock (internalHandlersLock)
-                    internalHandlers.Remove(handler);
-            }
-
-            public IEnumerable<object> AsEnumerable()
-            {
-                lock (internalHandlersLock)
-                    return internalHandlers.ToArray();
-            }
-        }
-
-        private readonly ConcurrentDictionary<Type, Handlers> subscriptions = new ConcurrentDictionary<Type, Handlers>();
+        private readonly ConcurrentDictionary<Type, object> subscriptions = new ConcurrentDictionary<Type, object>();
 
+        /// <inheritdoc />
         public void Publish<TEvent>(TEvent @event)
         {
             if (!subscriptions.TryGetValue(typeof(TEvent), out var handlers))
                 return;
-            foreach (var handler in handlers.AsEnumerable())
-                ((Action<TEvent>)handler)(@event);
+
+            ((Handlers<TEvent>) handlers).Handle(@event);
         }
 
-        public IDisposable Subscribe<TEvent>(Action<TEvent> eventHandler)
+        /// <inheritdoc />
+        public IDisposable Subscribe<TEvent>(Action<TEvent> handler)
         {
-            AddSubscription(eventHandler);
-            return GetCancelSubscriptionDelegate(eventHandler);
+            var handlers = (Handlers<TEvent>)subscriptions.GetOrAdd(typeof(TEvent), _ => new Handlers<TEvent>());
+            handlers.Add(handler);
+            return new Subscription<TEvent>(handlers, handler);
         }
 
-        private void AddSubscription<TEvent>(Action<TEvent> handler)
+        private sealed class Handlers<TEvent>
         {
-            subscriptions.AddOrUpdate(
-                typeof(TEvent),
-                new Handlers(handler),
-                (key, existingHandlers) =>
+            private readonly ILog log = LogProvider.For<Handlers<TEvent>>();
+            private readonly object mutex = new object();
+            private volatile List<Action<TEvent>> handlers = new List<Action<TEvent>>();
```

Comment:
`private volatile List<Action<TEvent>> handlers;` - you will eventually set this in `Add`.